File: apps/ios/ExpUI/Sources/Picker/PriorityPicker.swift
import SwiftUI

// EXP-1029 contract — the priority picker: contract `issuePriority` in its
// order, each by the priority glyph in its tone. The app hands in its
// `IssuePriority` option rows; the picker never owns the table.

public struct PriorityPickerOption: Identifiable, Hashable {
    public var id: String { value }
    /// Contract `issuePriority`.
    public let value: String
    public let label: String
    /// An `AppIcons` name.
    public let icon: String?
    public let colorHex: String?

    public init(value: String, label: String, icon: String? = nil, colorHex: String? = nil) {
        self.value = value
        self.label = label
        self.icon = icon
        self.colorHex = colorHex
    }
}

public struct PriorityPicker<Trigger: View>: View {
    public let options: [PriorityPickerOption]
    public let mode: PickerMode
    public let value: Set<String>
    public let onChange: (Set<String>) -> Void
    private let trigger: () -> Trigger

    public init(
        options: [PriorityPickerOption],
        mode: PickerMode = .single,
        value: Set<String>,
        onChange: @escaping (Set<String>) -> Void,
        @ViewBuilder trigger: @escaping () -> Trigger
    ) {
        self.options = options
        self.mode = mode
        self.value = value
        self.onChange = onChange
        self.trigger = trigger
    }

    public static func items(_ options: [PriorityPickerOption]) -> [PickerItem<String>] {
        options.map { option in
            PickerItem(
                value: option.value,
                label: option.label,
                icon: option.icon,
                color: option.colorHex.flatMap { Color(hex: $0) }
            )
        }
    }

    public var body: some View {
        GlassPicker(
            items: Self.items(options),
            mode: mode,
            value: value,
            onChange: onChange,
            title: "Priority",
            trigger: trigger
        )
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/components/picker/PriorityPicker.kt
package com.exponential.app.ui.components.picker

import androidx.compose.runtime.Composable
import com.exponential.app.domain.IssuePriority
import com.exponential.app.domain.priorityIcon
import com.exponential.app.ui.theme.priorityColor

/**
 * EXP-1029 contract, EXP-1021 implementation — the priority picker: contract `issuePriority` in its
 * order, each by the priority glyph in its tone. The app hands in its
 * priority option rows; the picker never owns the table.
 */
data class PriorityPickerOption(
    /** Contract `issuePriority`. */
    val value: String,
    val label: String,
    val colorHex: String? = null,
)

fun priorityPickerItems(options: List<PriorityPickerOption>): List<PickerItem<String>> =
    options.map { option ->
        // Glyph and tone are FUNCTIONS of the contract value (`priorityIcon` /
        // `priorityColor`), so a caller can never hand in a priority row that
        // draws differently from the same priority in a list.
        val priority = IssuePriority.fromWire(option.value)
        PickerItem(
            value = option.value,
            label = option.label,
            icon = priorityIcon(priority),
            color = priorityColor(priority),
        )
    }

@Composable
fun PriorityPicker(
    options: List<PriorityPickerOption>,
    value: Set<String>,
    onChange: (Set<String>) -> Unit,
    mode: PickerMode = PickerMode.Single,
    /** The sheet headline; the default names the picker. */
    title: String = "Priority",
    /**
     * EXP-1021: the sheet CONTROLLED by the caller, for a picker that is a
     * state machine rather than a chip (the issue screens open theirs from a
     * properties sheet that has already closed).
     */
    open: Boolean? = null,
    onOpenChange: ((Boolean) -> Unit)? = null,
    trigger: @Composable (open: () -> Unit) -> Unit = {},
) {
    Picker(
        items = priorityPickerItems(options),
        mode = mode,
        value = value,
        onChange = onChange,
        title = title,
        open = open,
        onOpenChange = onOpenChange,
        trigger = trigger,
    )
}
